feature: Add apt package to universe

cpuguy83 opened the pull request:
This packages is fairly similar in interface to the alpine package.

Some specific features of this over a standard ubuntu install:

1. Configured to cache packages
2. Stores apt cache in cache mounts
3. For the purposes of installing packages specified, it undoes some of
   the built-in apt configs that help keep the image smaller since these
   are not needed when using cache mounts and messes up the cache.
4. Some optimizations so that `apt-get update` is only run when the apt
   cache is older than 60 minutes (or non-existent).

cpuguy83 commented:
I cannot reproduce the linting error

TomChv commented on pkg/universe.dagger.io/ubuntu/test/test.bats:
```diff
@@ -0,0 +1,8 @@
+setup() {
+    load '../../bats_helpers'
```
`../../../../bats_helpers` since your package will be in `universe.dagger.io/x/<your email>`

TomChv commented on pkg/universe.dagger.io/ubuntu/ubuntu.cue:
```diff
@@ -0,0 +1,129 @@
+	}
+
+	_sources_file: {}
+	if sources != "" {
+		_sources_file: core.#WriteFile & {
+			input:    dagger.#Scratch
+			path:     "/sources.list"
+			contents: sources
+		}
+	}
```
Nit: Avoid if, prefer [disjunction](https://docs.dagger.io/1226/coding-style#favor-disjunctions-over-if-conditions)

TomChv commented on pkg/universe.dagger.io/ubuntu/ubuntu.cue:
```diff
@@ -0,0 +1,129 @@
+			docker.#Pull & {
+				source: "index.docker.io/ubuntu:\(version)"
+			},
+			if sources != "" {
+				docker.#Copy & {
+					contents: _sources_file.output
+					dest:     "/etc/apt/sources.list"
+				}
+			},
```
Nit: Avoid if, prefer [disjunction](https://docs.dagger.io/1226/coding-style#favor-disjunctions-over-if-conditions)

TomChv commented on pkg/universe.dagger.io/ubuntu/ubuntu.cue:
```diff
@@ -0,0 +1,129 @@
+							dest:     "/etc/apt/apt.conf.d/docker-gzip-indexes"
+						}
+						"keep-cache": {
+							type:     "fs"
```
Nit: `type: "fs"` is infered, you could remove it

TomChv reviewed (commented):
Thanks for your contribution.

Experimental package should be in `universe.dagger.io/x/<your email>`, you can find example [here](https://github.com/dagger/dagger/tree/main/pkg/universe.dagger.io/x).

I left nit comments to improve it.

Fix the linter, put the package in `universe.dagger.io` and we can merge that one after another pair of eyes look at your PR :D

cpuguy83 commented:
Updated.

- Moved to `x/<email>`
- Changed package from `ubuntu` to `apt`
- Now includes an `#Install` (instead of `#Build`) action which is a build step that can be inserted into a `docker.#Build` to install packages.

sipsma commented on pkg/universe.dagger.io/x/cpuguy83@gmail.com/apt/install.cue:
```diff
@@ -0,0 +1,115 @@
+						name: "/bin/sh"
+						args: ["-c", """
+                            set -e
+                            if [ -z "$(find ${DAGGER_UPDATED_PATH} -mmin -60)" ]; then 
```
1. This is neat
2. Do you think it would make sense to make the `60` configurable? In case someone wants to force a re-run of apt update? Not necessary if it's hard for whatever reason, just a thought in case it's easy.

cpuguy83 replied:
Added.

sipsma reviewed (commented):
Just quick nits and suggestions, looks awesome! I have a vague, unpleasant memory of trying to add all these types optimizations while writing a Dockerfile several years ago, great to see it will be so easy + reusable now.

cpuguy83 commented on pkg/universe.dagger.io/x/cpuguy83@gmail.com/apt/install.cue:
```diff
@@ -0,0 +1,115 @@
+)
+
+_#aptMounts: {
+	cache_prefix:     string
```
Done.

cpuguy83 commented:
Seems like there's some issues with gha cache.

sipsma commented:
> Seems like there's some issues with gha cache.

Yeah I'm taking a look at the Gitpod issue first and I'll look at the GHA one next. None of it has to do with your PR though

shykes reviewed (approved):
Pro-actively approving so we can merge as soon as tests pass. Thanks for the contribution Brian!

I think we should consider fast-tracking this to an official universe package too, given the API similarity to alpine... Seems relatively low risk.

sipsma commented:
@cpuguy83 the Gitpod issue has been fixed in main, once you rebase this should be good to go. I confirmed in the jaeger traces that the timeouts in the test were just 429's from GHA, so they should pass on retries.
